rfd/0222-access-list-templates.md

```diff
@@ -0,0 +1,258 @@
+- Product: @r0mant
+
+## What
+
+Add support for templated Access List that have system behaviors such as Teleport creating the required roles and then Teleport assigning those roles to members and owners upon creating an Access List.
+
+## Why
+
+Improves Access List usability especially for day one users. Templated Access List allows an admin to focus on users and what resources users should have access to. It removes the need for an admin to learn how to create roles and removes how roles have relation to an Access List because Teleport will do it for them.
+
+### User story: As an admin, I want to create an access list that require members to request for access and then grants short-term access to resources
+
+The template type to use for this case is `short-term`.
+
+Template type `short-term` represents an access list that utilizes JIT. Owners are reviewers. Members are requesters that are required to request access to resources and then upon approval are granted short-term access to requested Teleport resources.
+
+Admin will define what resources members will have access to by specifying the resource kinds and its labels and its the resource principals.
+
+### User story: As an admin, I want to create an accesss list that grants long-term access to resources for members
+
+The template type to use for this case is `long-term`.
+
+Template type `long-term` represents an access list that grants members standing access to Teleport resources. Owners will have no special purpose other than to audit.
+
+Admin will define what resources members will have access to by specifying the resource kinds and its labels and its the resource principals.
+
+This type of template is similar to how access list works now (non integrated types). Only difference is Teleport will create the necessary role for the admin.
+
+### Data Model
+
+#### Type Field
+
+In the current [AccessListSpec](https://github.com/gravitational/teleport/blob/bbb0f46b22ff88299908bef8dcf85d292aa379e1/api/proto/teleport/accesslist/v1/accesslist.proto#L75) model, there exists a field called `type`. We will introduce an additional type called `template` to indicate an access list used a template upon creation.
+
+```proto
+message AccessListSpec {
+  // ... other existing fields
+
+  // Existing type values: dynamic, scim, and static
+  // NEW value: "template"
+  string type = 12;
+}
+```
+
+#### Template Specification
+
+A new field `template_config` will be introduced in the current [AccessListSpec](https://github.com/gravitational/teleport/blob/bbb0f46b22ff88299908bef8dcf85d292aa379e1/api/proto/teleport/accesslist/v1/accesslist.proto#L75) model.
+
+```proto
+message AccessListSpec {
+  // ... other existing fields
+
+  AccessListTemplateConfig template_config = 13;
+}
+
+// AccessListTemplateConfig describes the template used.
+message AccessListTemplateConfig {
+  oneof template {
+    TemplateLongTerm long_term = 1;
+    TemplateShortTerm short_term = 2;
+  }
+}
```

Comment:
Would be good to check how good `oneof` integrates with our Terraform provider. Possibly it would be better Terraform UX to change it to have `type` field instead.

Reply:
it doesn't look too bad, here is how the fields are [documented](https://github.com/gravitational/teleport/blob/1fdc377456e00a7c5db2c0a9ecdf26e0d50a3fcb/docs/pages/reference/terraform-provider/data-sources/access_list.mdx#nested-schema-for-spectemplate_config) (generated file)

and here is how i crafted the example based on documentation: https://github.com/gravitational/teleport/pull/58297/files#diff-cf3663b2d5041c9e3d6c7c4c0e66a7882cb17fd6ae833cf41ab17b24ca9e511eR184